### PR TITLE
enforce complete type annotations

### PR DESCRIPTION
**Pull request by wochinge**
**Proposed changes**:
- fix https://github.com/RasaHQ/rasa/issues/8247

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by wochinge** [comment]
@m-vdb Do you think it would be an option to force-merge this **if** the only failing things are docstrings? Because of the reach of my changes it's quite a lot of failing docstring checks (they wouldn't fail for anyone else as `pydocstyle` only runs on diffs)

**Comment by m-vdb** [comment]
yeah I think that'd be ok 👍🏻

File: rasa/core/utils.py [hunk 1]
```diff
@@ -363,28 +364,6 @@ def read_endpoints_from_path(
     return AvailableEndpoints.read_endpoints(endpoints_config_path)
 
 
-# noinspection PyProtectedMember
-def set_default_subparser(parser, default_subparser) -> None:
```
Comment (m-vdb) [inline rasa/core/utils.py:367]: was this dead code?

**Review by m-vdb — APPROVED** [review]
⭐ nice!

File: rasa/cli/utils.py [hunk 2]
```diff
@@ -196,6 +196,6 @@ def payload_from_button_question(button_question: "Question") -> Text:
     return response
 
 
-def signal_handler(sig, frame) -> NoReturn:
+def signal_handler(sig: int, frame: Any) -> NoReturn:
```
[the remark this answers is not in the document]
Comment (wochinge) [reply rasa/cli/utils.py:199]: thanks, was looking for this one but coudn't find it.

File: rasa/core/training/interactive.py [hunk 3]
```diff
@@ -4,6 +4,7 @@
 import tempfile
 import textwrap
 import uuid
+from asyncio import AbstractEventLoop
```
[the remark this answers is not in the document]
Comment (wochinge) [reply rasa/core/training/interactive.py:7]: not really as we are importing asyncio already. Will use `asyncio.AbstractEventLoop` though

File: rasa/core/utils.py (hunk 1 above)
Comment (wochinge) [reply rasa/core/utils.py:367]: yes

File: Makefile [hunk 4]
```diff
@@ -118,7 +118,8 @@ types:
 	--disable-error-code return \
 	--disable-error-code has-type \
 	--disable-error-code func-returns-value \
-	--disallow-untyped-calls
+	--disallow-untyped-calls \
+	--disallow-incomplete-defs
```
[the remark this answers is not in the document]
Comment (wochinge) [reply Makefile]: moved the entire config to setup.cfg 👍🏻

File: setup.cfg [hunk 5]
```diff
@@ -32,8 +32,15 @@ docstring-convention = google
...
 warn_redundant_casts = True
 warn_unused_ignores = True
-# FIXME: more configuration using the --disable-error-code CLI argument (see types directive in Makefile)
+disallow_untyped_calls = True
+disallow_incomplete_defs = True
+# FIXME: working our way towards removing these
+	# see https://github.com/RasaHQ/rasa/pull/6470
+	# the list below is sorted by the number of errors for each error code, in decreasing order
+disable_error_code = arg-type, assignment, var-annotated, return-value, union-attr,
+    override, operator, attr-defined, index, misc, return, has-type, func-returns-value
```
Comment (m-vdb) [inline setup.cfg:46]: ❤️

File: setup.cfg [hunk 6]
```diff
@@ -32,8 +32,15 @@ docstring-convention = google
 [mypy]
 # some libraries do not have type hints; ideally we remove
 # this at some point but it will require a lot of efforts from us and library owners
+mypy_path=./stubs
```
Comment (m-vdb) [inline setup.cfg:35]:
```suggestion
mypy_path = ./stubs
```

**Comment by wochinge** [comment]
@m-vdb mypy is passing locally and the rest of the test is passing as well. One of our memory leaks checks failed, but I'd increase the value rather in a separate PR (definitely not related to type annotations)

**Comment by wochinge** [comment]
doing a PR for the memory leak test [here](https://github.com/RasaHQ/rasa/pull/8378)

**Comment by m-vdb** [comment]
@wochinge did the "Test CLI" step pass locally as well?

**Comment by wochinge** [comment]
yep, did that! Works

**Comment by m-vdb** [comment]
@wochinge what's up with [this test](https://github.com/RasaHQ/rasa/pull/8354/checks?check_run_id=2285878858)?

**Comment by wochinge** [comment]
Pretty sure it's flakey - will see if I can fix it but not related to this PR

**Comment by wochinge** [comment]
Anca fixed the failing test [here](https://github.com/RasaHQ/rasa/pull/8346/files) . Ready to merge then @m-vdb

**Pull request merged**